<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>mrIML • mrIML</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="mrIML">
<meta property="og:description" content="This package aims to enable users to build and interpret multivariate machine learning models harnessing the tidyverse (tidy model syntax in particular). This package builds off ideas from Gradient Forests (Ellis et al 2012), ecological genomic approaches (Fitzpatrick and Keller, 2014) and multi-response stacking algorithms (Xing et al 2019).">
<meta property="og:image" content="nfj1380.io/mrIML/reference/figures/logo.png">
<meta property="og:image:alt" content="Multivariate–multi-response–interpretable machine learning">
<meta name="twitter:card" content="summary">
<meta name="twitter:creator" content="@gustavoetal">
<meta name="twitter:site" content="@gustavoetal">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">mrIML</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Vignette_biosecurity.html">Swine Biosecurity working example</a>
    </li>
    <li>
      <a href="articles/Vignette_classification.html">Classification working example</a>
    </li>
    <li>
      <a href="articles/Vignette_regression.html">Regression working example</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="mriml-multivariate-multi-response-interpretable-machine-learning-">mrIML: Multivariate (multi-response) interpretable machine learning <a href="https://nfj1380.github.io/mrIML/index.html" class="external-link"><img src="reference/figures/logo.png" align="right" height="120"></a>
<a class="anchor" aria-label="anchor" href="#mriml-multivariate-multi-response-interpretable-machine-learning-"></a>
</h1></div>
<!-- badges: start -->

<p>This package aims to enable users to build and interpret multivariate machine learning models harnessing the tidyverse (tidy model syntax in particular). This package builds off ideas from Gradient Forests <a href="https://esajournals-onlinelibrary-wiley-com.prox.lib.ncsu.edu/doi/full/10.1890/11-0252.1" class="external-link">Ellis et al 2012</a>), ecological genomic approaches <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/ele.12376" class="external-link">Fitzpatrick and Keller, 2014</a> and multi-response stacking algorithms [Xing et al 2019].</p>
<p>This package can be of use for any multi-response machine learning problem, but was designed to handle data common to community ecology (site by species data) and ecological genomics (individual or population by SNP loci).</p>
<div class="section level2">
<h2 id="recent-mriml-publications">Recent mrIML publications<a class="anchor" aria-label="anchor" href="#recent-mriml-publications"></a>
</h2>
<ol style="list-style-type: decimal">
<li><p>Fountain-Jones, N. M., Kozakiewicz, C. P., Forester, B. R., Landguth, E. L., Carver, S., Charleston, M., Gagne, R. B., Greenwell, B., Kraberger, S., Trumbo, D. R., Mayer, M., Clark, N. J., &amp; Machado, G. (2021). MrIML: Multi-response interpretable machine learning to model genomic landscapes. Molecular Ecology Resources, 21, 2766– 2781. <a href="https://doi.org/10.1111/1755-0998.13495" class="external-link uri">https://doi.org/10.1111/1755-0998.13495</a></p></li>
<li><p>Sykes, A. L., Silva, G. S., Holtkamp, D. J., Mauch, B. W., Osemeke, O., Linhares, D. C.L., &amp; Machado, G. (2021). Interpretable machine learning applied to on-farm biosecurity and porcine reproductive and respiratory syndrome virus. Transboundary and Emerging Diseases, 00, 1– 15. <a href="https://doi.org/10.1111/tbed.14369" class="external-link uri">https://doi.org/10.1111/tbed.14369</a></p></li>
</ol>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install the stable version of the package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#install.packages("devtools")</span>
<span class="co">#devtools::install_github('nfj1380/mrIML')</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nfj1380.github.io/mrIML" class="external-link">mrIML</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p><strong>mrIML</strong> is designed to be used with a single function call or to be used in an ad-hoc fashion via individual function calls. In the following section we give an overview of the simple use case. For more on using each function see the <a href="https://nfj1380.github.io/mrIML/reference/index.html" class="external-link">function documentation</a>. The core functions for both regression and classification are: <a href="https://nfj1380.github.io/mrIML/reference/mrIMLpredicts.html" class="external-link"><code>mrIMLpredicts</code></a>, <a href="https://nfj1380.github.io/mrIML/reference/mrIMLperformance.html" class="external-link"><code>mrIMLperformance</code></a>, and <a href="https://nfj1380.github.io/mrIML/reference/mrInteractions.html" class="external-link"><code>mrInteractions</code></a>,for plotting and visualization <a href="https://nfj1380.github.io/mrIML/reference/mrVip.html" class="external-link"><code>mrVip</code></a>, <a href="https://nfj1380.github.io/mrIML/reference/mrFlashlight.html" class="external-link"><code>mrFlashlight</code></a>, and<a href="https://nfj1380.github.io/mrIML/reference/plot_vi.html" class="external-link"><code>plot_vi</code></a>. Estimating the interactions alone can be substantially computationally demanding depending on the number of outcomes you want to test. The first step to using the package is to load it as follows.</p>
</div>
<div class="section level2">
<h2 id="model-component">Model component<a class="anchor" aria-label="anchor" href="#model-component"></a>
</h2>
<p>Now all the data is loaded and ready to go we can formulate the model using tidymodel syntax. In this case we have binary data (SNP presence/absence at each loci) but the data could also be counts or continuous (the set_model argument would be “regression” instead of “classification”). The user can specify any model from the ‘tidymodel’ universe as ‘model 1’ (see <a href="https://www.tidymodels.org/find/" class="external-link uri">https://www.tidymodels.org/find/</a> for details). However, we have done most of our testing on random forests (rf) and glms (generalized linear models). Here we will specify a random forest classification model as the model applied to each response.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">model1</span> <span class="op">&lt;-</span> 
  <span class="fu">rand_forest</span><span class="op">(</span>trees <span class="op">=</span> <span class="fl">10</span>, mode <span class="op">=</span> <span class="st">"classification"</span>, 
              mtry <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>, 
              min_n <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co">#100 trees are set for brevity</span>
              <span class="fu">set_engine</span><span class="op">(</span><span class="st">"randomForest"</span><span class="op">)</span></code></pre></div>
<div class="section level3">
<h3 id="mrimlpredicts">
<a href="https://nfj1380.github.io/mrIML/reference/mrIMLpredicts.html" class="external-link"><code>mrIMLpredicts</code></a><a class="anchor" aria-label="anchor" href="#mrimlpredicts"></a>
</h3>
<p>This function represents the core functionality of the package and includes results reporting, plotting and optional saving. It requires a data frame of X t( the snp data for example) and Y represented by the covariates or features.</p>
<p>Load example data (cite) data from <a href="https://nfj1380.github.io/mrIML" class="external-link">mrIML</a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fData</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/filterRareCommon.html">filterRareCommon</a></span> <span class="op">(</span><span class="va">Responsedata</span>,
                           lower<span class="op">=</span><span class="fl">0.4</span>,
                           higher<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> 
<span class="va">data</span> <span class="op">&lt;-</span> <span class="va">fData</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="parallel-processing">Parallel processing<a class="anchor" aria-label="anchor" href="#parallel-processing"></a>
</h2>
<p>MrIML provides uses the flexible future apply functionality to set up multi-core processing. In the example below, we set up a cluster using 4 cores. If you don’t set up a cluster, the default settings will be used and the analysis will run sequentially.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># detectCores() #check how many cores you have available. We suggest keeping one core free for internet browsing etc.</span>

<span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span>

<span class="fu">plan</span><span class="op">(</span><span class="va">cluster</span>,
     workers<span class="op">=</span><span class="va">cl</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Define set of features</span>
<span class="va">fData</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/filterRareCommon.html">filterRareCommon</a></span> <span class="op">(</span><span class="va">Responsedata</span>,
                           lower<span class="op">=</span><span class="fl">0.4</span>,
                           higher<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> 
<span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">fData</span> <span class="co">#For simplicity when comparing</span>
<span class="co">#Define set the outcomes of interest</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">Features</span><span class="op">)</span> 
<span class="co">#&gt; 'data.frame':    20 obs. of  19 variables:</span>
<span class="co">#&gt;  $ Grassland       : num  0.07 0.0677 0.1845 0.0981 0.1578 ...</span>
<span class="co">#&gt;  $ Shrub.Scrub     : num  0.557 0.767 0.524 0.786 0.842 ...</span>
<span class="co">#&gt;  $ Forest          : num  0.01072 0.030588 0.008615 0.000662 0.000616 ...</span>
<span class="co">#&gt;  $ HighlyDev       : num  0 0 0.00225 0 0 ...</span>
<span class="co">#&gt;  $ Urban           : num  0 0 0.00159 0 0 ...</span>
<span class="co">#&gt;  $ Suburban        : num  0.00357 0.13268 0.01325 0.00119 0 ...</span>
<span class="co">#&gt;  $ Exurban         : num  0.00622 0.03019 0 0.01906 0 ...</span>
<span class="co">#&gt;  $ Altered         : num  0.441 0.182 0.114 0.12 0 ...</span>
<span class="co">#&gt;  $ Distance        : num  1.321 0.492 3.231 5.629 4.739 ...</span>
<span class="co">#&gt;  $ Latitude        : num  33.8 33.8 33.8 33.8 33.8 ...</span>
<span class="co">#&gt;  $ Longitude       : num  -118 -118 -118 -118 -118 ...</span>
<span class="co">#&gt;  $ Age             : int  3 0 3 2 3 3 2 3 3 3 ...</span>
<span class="co">#&gt;  $ Sex             : int  1 1 1 1 0 0 0 1 1 1 ...</span>
<span class="co">#&gt;  $ Relatedness.PCO1: num  -0.1194 -0.0389 -0.1618 -0.1811 -0.1564 ...</span>
<span class="co">#&gt;  $ Relatedness.PCO2: num  -0.1947 -0.0525 -0.321 -0.0827 0.1 ...</span>
<span class="co">#&gt;  $ Relatedness.PCO3: num  -0.191 -0.0874 0.0541 -0.0627 -0.0111 ...</span>
<span class="co">#&gt;  $ Relatedness.PCO4: num  0.1117 0.2422 0.0974 0.2129 0.2259 ...</span>
<span class="co">#&gt;  $ Relatedness.PCO5: num  0.06405 0.0706 0.03514 -0.00084 0.0894 ...</span>
<span class="co">#&gt;  $ Relatedness.PCO6: num  -0.0432 0.0683 -0.0805 0.2247 -0.055 ...</span>
<span class="co">#Remove NAs from the feature/predictor data.</span>
<span class="va">FeaturesnoNA</span><span class="op">&lt;-</span><span class="va">Features</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html" class="external-link">complete.cases</a></span><span class="op">(</span><span class="va">Features</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="va">FeaturesnoNA</span> <span class="co">#For simplicity</span>
<span class="co">#For more efficient testing for interactions (more variables more interacting pairs)</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="va">FeaturesnoNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span> <span class="co">#Three features only</span>


<span class="va">yhats</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">X</span>, <span class="co">#Features/predictors </span>
                       Y<span class="op">=</span><span class="va">Y</span>, <span class="co">#Response data</span>
                       Model<span class="op">=</span><span class="va">model1</span>, <span class="co">#Specify your model</span>
                       balance_data<span class="op">=</span><span class="st">'no'</span>, <span class="co">#Chose how to balance your data </span>
                       mode<span class="op">=</span><span class="st">'classification'</span>, <span class="co">#Chose your mode (classification versus regression)</span>
                       seed <span class="op">=</span> <span class="fl">120</span><span class="op">)</span> <span class="co">#Set seed</span>

<span class="va">ModelPerf</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span>yhats<span class="op">=</span><span class="va">yhats</span>,
                              Model<span class="op">=</span><span class="va">model1</span>,
                              Y<span class="op">=</span><span class="va">Y</span>, mode<span class="op">=</span><span class="st">'classification'</span><span class="op">)</span>
<span class="va">ModelPerf</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co">#Predictive performance for individual responses </span>
<span class="co">#&gt;    response  model_name           roc_AUC                mcc       sensitivity       specificity</span>
<span class="co">#&gt; 1   env_131 rand_forest                 1  0.666666666666667 0.666666666666667                 1</span>
<span class="co">#&gt; 2   env_163 rand_forest              0.75               &lt;NA&gt;                 0                 1</span>
<span class="co">#&gt; 3   env_164 rand_forest                 1                  1                 1                 1</span>
<span class="co">#&gt; 4   env_167 rand_forest 0.583333333333333 -0.408248290463863 0.666666666666667                 0</span>
<span class="co">#&gt; 5   env_169 rand_forest                 1  0.666666666666667 0.666666666666667                 1</span>
<span class="co">#&gt; 6   env_212 rand_forest                 1                  1                 1                 1</span>
<span class="co">#&gt; 7    env_23 rand_forest             0.875  0.408248290463863                 1               0.5</span>
<span class="co">#&gt; 8    env_24 rand_forest              0.25 -0.166666666666667               0.5 0.333333333333333</span>
<span class="co">#&gt; 9    env_41 rand_forest 0.583333333333333  0.166666666666667 0.666666666666667               0.5</span>
<span class="co">#&gt; 10   env_47 rand_forest                 1  0.666666666666667                 1 0.666666666666667</span>
<span class="co">#&gt; 11   env_59 rand_forest                 1  0.666666666666667 0.666666666666667                 1</span>
<span class="co">#&gt; 12    env_8 rand_forest 0.666666666666667               &lt;NA&gt;                 1                 0</span>
<span class="co">#&gt; 13   env_84 rand_forest                 1  0.666666666666667 0.666666666666667                 1</span>
<span class="co">#&gt; 14   env_85 rand_forest                 1  0.666666666666667 0.666666666666667                 1</span>
<span class="co">#&gt; 15   env_86 rand_forest               0.5               &lt;NA&gt;                 0                 1</span>
<span class="co">#&gt; 16  pol_105 rand_forest             0.875  0.612372435695795              0.75                 1</span>
<span class="co">#&gt; 17  pol_108 rand_forest              0.75  0.408248290463863               0.5                 1</span>
<span class="co">#&gt; 18  pol_111 rand_forest                 1                  1                 1                 1</span>
<span class="co">#&gt; 19  pol_117 rand_forest 0.583333333333333  0.166666666666667 0.666666666666667               0.5</span>
<span class="co">#&gt; 20  pol_132 rand_forest                 1  0.666666666666667                 1 0.666666666666667</span>
<span class="co">#&gt; 21  pol_159 rand_forest              0.25 -0.166666666666667               0.5 0.333333333333333</span>
<span class="co">#&gt; 22  pol_258 rand_forest              0.25 -0.166666666666667               0.5 0.333333333333333</span>
<span class="co">#&gt; 23   pol_30 rand_forest                 1  0.666666666666667                 1 0.666666666666667</span>
<span class="co">#&gt; 24  pol_340 rand_forest 0.416666666666667 -0.166666666666667 0.333333333333333               0.5</span>
<span class="co">#&gt; 25  pol_353 rand_forest               0.5               &lt;NA&gt;                 0                 1</span>
<span class="co">#&gt; 26  pol_366 rand_forest                 1  0.666666666666667 0.666666666666667                 1</span>
<span class="co">#&gt; 27   pol_87 rand_forest                 1  0.666666666666667                 1 0.666666666666667</span>
<span class="co">#&gt; 28   pol_88 rand_forest                 1  0.666666666666667                 1 0.666666666666667</span>
<span class="co">#&gt; 29   pol_89 rand_forest                 1  0.666666666666667                 1 0.666666666666667</span>
<span class="co">#&gt;           prevalence</span>
<span class="co">#&gt; 1  0.421052631578947</span>
<span class="co">#&gt; 2  0.631578947368421</span>
<span class="co">#&gt; 3  0.421052631578947</span>
<span class="co">#&gt; 4  0.421052631578947</span>
<span class="co">#&gt; 5  0.421052631578947</span>
<span class="co">#&gt; 6  0.684210526315789</span>
<span class="co">#&gt; 7  0.631578947368421</span>
<span class="co">#&gt; 8  0.421052631578947</span>
<span class="co">#&gt; 9  0.473684210526316</span>
<span class="co">#&gt; 10 0.473684210526316</span>
<span class="co">#&gt; 11 0.421052631578947</span>
<span class="co">#&gt; 12 0.473684210526316</span>
<span class="co">#&gt; 13 0.421052631578947</span>
<span class="co">#&gt; 14 0.421052631578947</span>
<span class="co">#&gt; 15 0.421052631578947</span>
<span class="co">#&gt; 16 0.473684210526316</span>
<span class="co">#&gt; 17 0.421052631578947</span>
<span class="co">#&gt; 18 0.421052631578947</span>
<span class="co">#&gt; 19 0.473684210526316</span>
<span class="co">#&gt; 20 0.473684210526316</span>
<span class="co">#&gt; 21 0.473684210526316</span>
<span class="co">#&gt; 22 0.473684210526316</span>
<span class="co">#&gt; 23 0.473684210526316</span>
<span class="co">#&gt; 24 0.421052631578947</span>
<span class="co">#&gt; 25 0.421052631578947</span>
<span class="co">#&gt; 26 0.421052631578947</span>
<span class="co">#&gt; 27 0.473684210526316</span>
<span class="co">#&gt; 28 0.473684210526316</span>
<span class="co">#&gt; 29 0.473684210526316</span>
<span class="va">ModelPerf</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="co">#Overall predictive performance. r2 for regression and MCC for classification</span>
<span class="co">#&gt; [1] 0.7873563</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">VI</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mrVip.html">mrVip</a></span><span class="op">(</span><span class="va">yhats</span>, X<span class="op">=</span><span class="va">X</span><span class="op">)</span> 
<span class="fu"><a href="reference/plot_vi.html">plot_vi</a></span><span class="op">(</span>VI<span class="op">=</span><span class="va">VI</span>,
        X<span class="op">=</span><span class="va">X</span>,
        Y<span class="op">=</span><span class="va">Y</span>,
        modelPerf<span class="op">=</span><span class="va">ModelPerf</span>, 
       cutoff<span class="op">=</span> <span class="fl">0</span>, mode<span class="op">=</span><span class="st">'classification'</span><span class="op">)</span> <span class="co">#The cutoff reduces the number of individual models printed in the second plot. </span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-8-1.png" width="100%"></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Press [enter] to plot individual variable importance summaries</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-8-2.png" width="100%"></p>
</div>
<div class="section level2">
<h2 id="effect-of-a-feature-on-genetic-change">Effect of a feature on genetic change<a class="anchor" aria-label="anchor" href="#effect-of-a-feature-on-genetic-change"></a>
</h2>
<p>We also wrap some flashlight functionality to visualize the marginal (i.e. partial dependencies) or conditional (accumulated local effects) effect of a feature on genetic change. Partial dependencies take longer to calculate and are more sensitive to correlated features</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flashlightObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mrFlashlight.html">mrFlashlight</a></span><span class="op">(</span><span class="va">yhats</span>,
                              X<span class="op">=</span><span class="va">X</span>,
                              Y<span class="op">=</span><span class="va">Y</span>,
                              response <span class="op">=</span> <span class="st">"single"</span>,
                              index<span class="op">=</span><span class="fl">1</span>,
                              mode<span class="op">=</span><span class="st">'classification'</span><span class="op">)</span>

<span class="co">#plot prediction scatter for all responses. Gets busy with </span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">light_scatter</span><span class="op">(</span><span class="va">flashlightObj</span>,
                   v <span class="op">=</span> <span class="st">"Forest"</span>,
                   type <span class="op">=</span> <span class="st">"predicted"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-9-1.png" width="100%"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#plots everything on one plot (partial dependency, ALE, scatter)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">light_effects</span><span class="op">(</span><span class="va">flashlightObj</span>,
                   v <span class="op">=</span> <span class="st">"Grassland"</span><span class="op">)</span>,
                   use <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-9-2.png" width="100%"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">#profileData_pd &lt;- light_profile(flashlightObj,  v = "Grassland")</span>

<span class="co">#mrProfileplot(profileData_pd , sdthresh =0.05) #sdthresh removes responses from the first plot that do not vary with the feature</span>

<span class="va">profileData_ale</span> <span class="op">&lt;-</span> <span class="fu">light_profile</span><span class="op">(</span><span class="va">flashlightObj</span>,
                                 v <span class="op">=</span> <span class="st">"Grassland"</span>,
                                 type <span class="op">=</span> <span class="st">"ale"</span><span class="op">)</span> <span class="co">#accumulated local effects</span>

<span class="fu"><a href="reference/mrProfileplot.html">mrProfileplot</a></span><span class="op">(</span><span class="va">profileData_ale</span>,
              sdthresh <span class="op">=</span><span class="fl">0.01</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-9-3.png" width="100%"></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  Press [enter] to continue to the global summary plot</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-9-4.png" width="100%"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#the second plot is the cumulative turnover function</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="interacting-predictors-or-features">Interacting predictors or features<a class="anchor" aria-label="anchor" href="#interacting-predictors-or-features"></a>
</h2>
<p>Finally, we can assess how features interact overall to shape genetic change. Be warned this is memory intensive. Future updates to this package will enable users to visualize these interactions and explore them in more detail using 2D ALE plots for example.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#interactions &lt;-mrInteractions(yhats,X,Y,mod='classification') #this is computationally intensive so multicores are needed. If stopped prematurely - have to reload things</span>
<span class="co">#mrPlot_interactions(interactions,X,Y,top_ranking = 2,top_response=2)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Xing, L, Lesperance, ML and Zhang, X (2020). Simultaneous prediction of multiple outcomes using revised stacking algorithms. Bioinformatics, 36, 65-72. <a href="doi:10.1093/bioinformatics/btz531" class="uri">doi:10.1093/bioinformatics/btz531</a>.</p>
<p>Fitzpatrick, M.C. &amp; Keller, S.R. (2015) Ecological genomics meets community-level modelling of biodiversity: mapping the genomic landscape of current and future environmental adaptation. Ecology Letters 18, 1–16.doi.org/10.1111/ele.12376</p>
<p>Ellis, N., Smith, S.J. and Pitcher, C.R. (2012), Gradient forests: calculating importance gradients on physical predictors. Ecology, 93: 156-168. <a href="doi:10.1890/11-0252.1" class="uri">doi:10.1890/11-0252.1</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/nfj1380/mrIML/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/nfj1380/mrIML/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing mrIML</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Nick Fountain-Jones <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9248-8493" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Nicholas Clark <br><small class="roles"> Author </small>  </li>
<li>Chris Kozakiewicz <br><small class="roles"> Author </small>  </li>
<li>
<a href="https://machado-lab.github.io/" class="external-link">Gustavo Machado</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-7552-6144" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/nfj1380/mrIML/actions" class="external-link"><img src="https://github.com/nfj1380/mrIML/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Fountain-Jones, Nicholas Clark, Chris Kozakiewicz, <a href="https://machado-lab.github.io/" class="external-link">Gustavo Machado</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
