<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>mrIML • mrIML</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="mrIML">
<meta property="og:description" content="This package aims to enable users to build and interpret multivariate machine learning models harnessing the tidyverse (tidy model syntax in particular). This package builds off ideas from Gradient Forests (Ellis et al 2012), ecological genomic approaches (Fitzpatrick and Keller, 2014) and multi-response stacking algorithms (Xing et al 2019).">
<meta property="og:image" content="nfj1380.io/mrIML/reference/figures/logo.png">
<meta property="og:image:alt" content="Multivariate–multi-response–interpretable machine learning">
<meta name="twitter:card" content="summary">
<meta name="twitter:creator" content="@gustavoetal">
<meta name="twitter:site" content="@gustavoetal">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">mrIML</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Vignette_LandscapeGenetics.html">Classification working example</a>
    </li>
    <li>
      <a href="articles/Vignette_regression.html">Regression working example</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
<li>
  <a href="reference/index.html">Functions</a>
</li>
<li>
  <a href="https://github.com/nfj1380/mrIML/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="mriml-multivariatemulti-responseinterpretable-machine-learning-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#mriml-multivariatemulti-responseinterpretable-machine-learning-" class="anchor"></a>mrIML: Multivariate–multi-response–interpretable machine learning <img src="reference/figures/logo.png" align="right" alt="" width="150">
</h1></div>
<p>Author: <strong>Nick Fountain-Jones</strong> -affiliation: <strong>University of Tasmania</strong></p>
<p>Author: <strong>Gustavo Machado</strong> -affiliation: <strong>North Carolina State University</strong> -Lab page: <a href="https://machado-lab.github.io/" class="uri">https://machado-lab.github.io/</a></p>
<!-- badges: start -->
<p><img src="https://img.shields.io/github/r-package/v/nfj1380/mrIML?logo=github&amp;logoColor=%2300ff37&amp;style=flat-square" alt="GitHub R package version"><img src="https://img.shields.io/github/contributors/nfj1380/mrIML?style=flat-square" alt="GitHub contributors"><img src="https://img.shields.io/github/last-commit/nfj1380/mrIML?style=flat-square" alt="GitHub last commit"></p>
<!-- badges: end -->
<p>This package aims to enable users to build and interpret multivariate machine learning models harnessing the tidyverse (tidy model syntax in particular). This package builds off ideas from Gradient Forests (Ellis et al 2012), ecological genomic approaches (Fitzpatrick and Keller, 2014) and multi-response stacking algorithms (Xing et al 2019).</p>
<p>This package can be of use for any multi-response machine learning problem, but was designed to handle data common to community ecology (site by species data) and ecological genomics (individual or population by SNP loci).</p>
<div id="-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#-installation" class="anchor"></a><i class="fas fa-cog"></i> Installation</h2>
<p>The current developmental version can be installed with</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">'nfj1380/mrIML'</span><span class="op">)</span></code></pre></div>
<div id="-dealing-with-errors" class="section level3">
<h3 class="hasAnchor">
<a href="#-dealing-with-errors" class="anchor"></a><i class="fas fa-skull-crossbones"></i> Dealing with Errors</h3>
<p>There are automatic checks for <a href="https://travis-ci.org/github//mrIML/branches"><strong>mrIML</strong></a>. However, that only checks for Linux and <strong>mrIML</strong> is built on Windows.</p>
</div>
</div>
<div id="-quick-start" class="section level2">
<h2 class="hasAnchor">
<a href="#-quick-start" class="anchor"></a><i class="fas fa-star"></i> Quick start</h2>
<p><strong>mrIML</strong> is designed to be used with a single function call or to be used in an ad-hoc fashion via individual function calls. In the following section we give an overview of the simple use case. For more on using each function see the <a href="https://nfj1380.github.io/mrIML/reference/index.html">function documentation</a>. The core functions for both regression and classification are: <a href="https://nfj1380.github.io/mrIML/reference/mrIMLpredicts.html"><code>mrIMLpredicts</code></a>, <a href="https://nfj1380.github.io/mrIML/reference/mrIMLperformance.html"><code>mrIMLperformance</code></a>, and <a href="https://nfj1380.github.io/mrIML/reference/mrInteractions.html"><code>mrInteractions</code></a>,for plotting and visualization <a href="https://nfj1380.github.io/mrIML/reference/mrVip.html"><code>mrVip</code></a>, <a href="https://nfj1380.github.io/mrIML/reference/mrFlashlight.html"><code>mrFlashlight</code></a>, and<a href="https://nfj1380.github.io/mrIML/reference/plot_vi.html"><code>plot_vi</code></a>. Estimating the interactions alone can be substantially computationally demanding depending on the number of outcomes you want to test. The first step to using the package is to load it as follows.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://nfj1380.github.io/mrIML">mrIML</a></span><span class="op">)</span>

<span class="co">#other package needed:</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/koalaverse/vip/">vip</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidymodels.tidymodels.org">tidymodels</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/">randomForest</a></span><span class="op">)</span>;  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/topepo/caret/">caret</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/gbm-developers/gbm">gbm</a></span><span class="op">)</span>;
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>;<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">doParallel</span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tidymodels/themis">themis</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sjmgarnier/viridis">viridis</a></span><span class="op">)</span>;
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sfirke/janitor">janitor</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/hrbrmstr/hrbrthemes">hrbrthemes</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dmlc/xgboost">xgboost</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://cran.r-project.org">vegan</a></span><span class="op">)</span>;<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mayer79/flashlight">flashlight</a></span><span class="op">)</span>;
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel">ggrepel</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://parsnip.tidymodels.org">parsnip</a></span><span class="op">)</span>;<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rsample.tidymodels.org">rsample</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tidymodels/workflows">workflows</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="-model-component" class="section level2">
<h2 class="hasAnchor">
<a href="#-model-component" class="anchor"></a><i class="fas fa-cog"></i> Model component</h2>
<p>Now all the data is loaded and ready to go we can formulate the model using tidymodel syntax. In this case we have binary data (SNP presence/absence at each loci) but the data could also be counts or continuous (the set_model argument would be “regression” instead of “classification”). The user can specify any model from the tidymodel universe as ‘model 1’ (see <a href="https://www.tidymodels.org/find/" class="uri">https://www.tidymodels.org/find/</a> for details). However, we have done most of our testing on random forests (rf), xgr boost and glms (generalized linear models). Here we will specify a random forest classification model as the model applied to each response.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">model1</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://parsnip.tidymodels.org/reference/rand_forest.html">rand_forest</a></span><span class="op">(</span>trees <span class="op">=</span> <span class="fl">100</span>, mode <span class="op">=</span> <span class="st">"classification"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co">#100 trees are set for brevity</span>
  <span class="fu"><a href="https://parsnip.tidymodels.org/reference/set_engine.html">set_engine</a></span><span class="op">(</span><span class="st">"ranger"</span>, importance <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"impurity"</span>,<span class="st">"impurity_corrected"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span><span class="co"># select the engine/package that underlies the model</span>
  <span class="fu"><a href="https://parsnip.tidymodels.org/reference/set_args.html">set_mode</a></span><span class="op">(</span><span class="st">"classification"</span><span class="op">)</span><span class="co"># choose either the continuous "regression" or binary "classification" mode</span></code></pre></div>
</div>
<div id="-mrimlpredicts" class="section level2">
<h2 class="hasAnchor">
<a href="#-mrimlpredicts" class="anchor"></a><i class="fas fa-toolbo"></i> <a href="https://nfj1380.github.io/mrIML/reference/mrIMLpredicts.html">mrIMLpredicts</a>
</h2>
<p>This function represents the core functionality of the package and includes results reporting, plotting and optional saving. It requires a data frame of X the loci or snip data for exampel and Y represented by the covariates</p>
<p>Load example data (cite) data from <a href="https://nfj1380.github.io/mrIML">mrIML</a>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> gfData[<span class="dv">1</span><span class="sc">:</span><span class="dv">80</span>]</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>        x     y bio_1 bio_10 bio_15 bio_18 bio_2 bio_7 elevation     MEM<span class="fl">.1</span>      MEM<span class="fl">.2</span>      MEM<span class="fl">.3</span>       MEM<span class="fl">.4</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="sc">-</span><span class="fl">101.57</span> <span class="fl">49.14</span>    <span class="dv">31</span>    <span class="dv">182</span>     <span class="dv">63</span>    <span class="dv">189</span>   <span class="dv">131</span>   <span class="dv">489</span>       <span class="dv">498</span>  <span class="fl">24.55053</span>   <span class="fl">7.444415</span>  <span class="sc">-</span><span class="fl">9.018635</span>  <span class="fl">0.06879434</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> <span class="sc">-</span><span class="fl">147.89</span> <span class="fl">69.10</span>  <span class="sc">-</span><span class="dv">107</span>     <span class="dv">85</span>     <span class="dv">60</span>     <span class="dv">89</span>   <span class="dv">110</span>   <span class="dv">507</span>       <span class="dv">400</span> <span class="sc">-</span><span class="fl">14.18000</span> <span class="sc">-</span><span class="fl">22.945314</span> <span class="sc">-</span><span class="fl">11.845836</span> <span class="fl">21.59719769</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> <span class="sc">-</span><span class="fl">148.51</span> <span class="fl">63.39</span>   <span class="sc">-</span><span class="dv">52</span>     <span class="dv">90</span>     <span class="dv">63</span>    <span class="dv">196</span>   <span class="dv">102</span>   <span class="dv">399</span>      <span class="dv">1040</span> <span class="sc">-</span><span class="fl">15.77654</span> <span class="sc">-</span><span class="fl">27.420568</span> <span class="sc">-</span><span class="fl">13.038813</span> <span class="fl">14.65433486</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>  <span class="sc">-</span><span class="fl">98.35</span> <span class="fl">54.51</span>   <span class="sc">-</span><span class="dv">12</span>    <span class="dv">159</span>     <span class="dv">52</span>    <span class="dv">204</span>   <span class="dv">103</span>   <span class="dv">510</span>       <span class="dv">213</span>  <span class="fl">20.73482</span>   <span class="fl">8.715802</span> <span class="sc">-</span><span class="fl">19.187171</span> <span class="sc">-</span><span class="fl">5.13771830</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Define set of features</span>
<span class="va">FeaturesnoNA</span><span class="op">&lt;-</span><span class="va">Features</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html">complete.cases</a></span><span class="op">(</span><span class="va">Features</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">FeaturesnoNA</span> <span class="co">#for simplicity</span>
<span class="co"># Define set the outcomes of interst</span>
<span class="va">fData</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/filterRareCommon.html">filterRareCommon</a></span> <span class="op">(</span><span class="va">Responsedata</span>, lower<span class="op">=</span><span class="fl">0.4</span>, higher<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> 
<span class="va">X</span> <span class="op">&lt;-</span> <span class="va">fData</span> <span class="co">#</span>

<span class="va">yhats</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">X</span>,Y<span class="op">=</span><span class="va">Y</span>, model1<span class="op">=</span><span class="va">model1</span>, balance_data<span class="op">=</span><span class="st">'no'</span>, mod<span class="op">=</span><span class="st">'classification'</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#save(yhats, file='logreg_model')</span>
<span class="va">ModelPerf</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats</span>, <span class="va">model1</span>, X<span class="op">=</span><span class="va">X</span><span class="op">)</span> <span class="co">#</span>
<span class="va">ModelPerf</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></code></pre></div>
</div>
<div id="ploting" class="section level2">
<h2 class="hasAnchor">
<a href="#ploting" class="anchor"></a>Ploting</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">VI</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mrVip.html">mrVip</a></span><span class="op">(</span><span class="va">yhats</span>, Y<span class="op">=</span><span class="va">Y</span><span class="op">)</span> 
<span class="fu"><a href="reference/plot_vi.html">plot_vi</a></span><span class="op">(</span>VI<span class="op">=</span><span class="va">VI</span>,  X<span class="op">=</span><span class="va">X</span>,Y<span class="op">=</span><span class="va">Y</span>, modelPerf<span class="op">=</span><span class="va">ModelPerf</span>, cutoff<span class="op">=</span> <span class="fl">0</span>, plot.pca<span class="op">=</span><span class="st">'yes'</span><span class="op">)</span> <span class="co">#the cutoff reduces the number of individual models printed in the second plot. </span></code></pre></div>
<p><img src="reference/figures/fig1.png"></p>
</div>
<div id="effect-of-a-feature-on-genetic-change" class="section level2">
<h2 class="hasAnchor">
<a href="#effect-of-a-feature-on-genetic-change" class="anchor"></a>Effect of a feature on genetic change</h2>
<p>We also wrap some flashlight functionality to visualize the marginal (i.e. partial dependencies) or conditional (accumulated local effects) effect of a feature on genetic change. Partial dependencies take longer to calculate and are more sensitive to correlated features</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">flashlightObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mrFlashlight.html">mrFlashlight</a></span><span class="op">(</span><span class="va">yhats</span>, <span class="va">X</span>, <span class="va">Y</span>, response <span class="op">=</span> <span class="st">"multi"</span>, model<span class="op">=</span><span class="st">'classification'</span><span class="op">)</span>

<span class="co">#plot prediction scatter for all responses. Gets busy with </span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/flashlight/man/light_scatter.html">light_scatter</a></span><span class="op">(</span><span class="va">flashlightObj</span>, v <span class="op">=</span> <span class="st">"Forest"</span>, type <span class="op">=</span> <span class="st">"predicted"</span><span class="op">)</span><span class="op">)</span>

<span class="co">#plots everything on one plot (partial dependency, ALE, scatter)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/flashlight/man/light_effects.html">light_effects</a></span><span class="op">(</span><span class="va">flashlightObj</span>, v <span class="op">=</span> <span class="st">"Grassland"</span><span class="op">)</span>, use <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span>


<span class="co">#profileData_pd &lt;- light_profile(flashlightObj,  v = "Grassland")</span>

<span class="co">#mrProfileplot(profileData_pd , sdthresh =0.05) #sdthresh removes responses from the first plot that do not vary with the feature</span>

<span class="va">profileData_ale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/flashlight/man/light_profile.html">light_profile</a></span><span class="op">(</span><span class="va">flashlightObj</span>, v <span class="op">=</span> <span class="st">"Grassland"</span>, type <span class="op">=</span> <span class="st">"ale"</span><span class="op">)</span> <span class="co">#acumulated local effects</span>

<span class="fu"><a href="reference/mrProfileplot.html">mrProfileplot</a></span><span class="op">(</span><span class="va">profileData_ale</span> , sdthresh <span class="op">=</span><span class="fl">0.01</span><span class="op">)</span>
<span class="co">#the second plot is the cumulative turnover function</span></code></pre></div>
</div>
<div id="features-interact" class="section level2">
<h2 class="hasAnchor">
<a href="#features-interact" class="anchor"></a>Features interact</h2>
<p>Finally, we can assess how features interact overall to shape genetic change. Be warned this is memory intensive. Future updates to this package will enable users to visualize these interactions and explore them in more detail using 2D ALE plots for example.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">interactions</span> <span class="op">&lt;-</span><span class="fu"><a href="reference/mrInteractions.html">mrInteractions</a></span><span class="op">(</span><span class="va">yhats</span>, <span class="va">X</span>, <span class="va">Y</span>,  mod<span class="op">=</span><span class="st">'classification'</span><span class="op">)</span> <span class="co">#this is computationally intensive so multicores are needed. If stopped prematurely - have to reload things</span>

<span class="fu"><a href="reference/mrPlot_interactions.html">mrPlot_interactions</a></span><span class="op">(</span><span class="va">interactions</span>, <span class="va">X</span>,<span class="va">Y</span>, top_ranking <span class="op">=</span> <span class="fl">2</span>, top_response<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
</div>
<div id="-references" class="section level2">
<h2 class="hasAnchor">
<a href="#-references" class="anchor"></a><i class="fas fa-pen-square"></i> References</h2>
<p>Xing, L, Lesperance, ML and Zhang, X (2020). Simultaneous prediction of multiple outcomes using revised stacking algorithms. Bioinformatics, 36, 65-72.</p>
<p>Fitzpatrick, M.C. &amp; Keller, S.R. (2015) Ecological genomics meets community-level modelling of biodiversity: mapping the genomic landscape of current and future environmental adaptation. Ecology Letters 18, 1–16.</p>
<p>Ellis, N., Smith, S.J. and Pitcher, C.R. (2012), Gradient forests: calculating importance gradients on physical predictors. Ecology, 93: 156-168. <a href="doi:10.1890/11-0252.1" class="uri">doi:10.1890/11-0252.1</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/nfj1380/mrIML/">https://​github.com/​nfj1380/​mrIML/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/nfj1380/mrIML/issues">https://​github.com/​nfj1380/​mrIML/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Nick Fountain-Jones <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9248-8493" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Nicholas Clark <br><small class="roles"> Author </small>  </li>
<li>Chris Kozakiewicz <br><small class="roles"> Author </small>  </li>
<li>
<a href="https://gustavoetal.org/">Gustavo Machado</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-7552-6144" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Nick Fountain-Jones, Nicholas Clark, Chris Kozakiewicz, <a href="https://gustavoetal.org/">Gustavo Machado</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
