<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Graphical network model (GNM) • mrIML</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Graphical network model (GNM)">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="inverse" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">mrIML</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fas fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Common%20errors%20in%20mrIML.html">Common errors in mrIML</a></li>
    <li><a class="dropdown-item" href="../articles/Swine%20biosecurity.html">Swine Biosecurity working example</a></li>
    <li><a class="dropdown-item" href="../articles/Landscape%20genetics%20(classification).html">Classification working example</a></li>
    <li><a class="dropdown-item" href="../articles/Regression.html">Regression working example</a></li>
    <li><a class="dropdown-item" href="../articles/Landscape%20genetics.html">Landscape genetics</a></li>
    <li><a class="dropdown-item" href="../articles/Graphical%20network%20model%20(GNM).html">Graphical network model (GNM)</a></li>
    <li><a class="dropdown-item" href="../articles/Microbiome%20data%20curation.html">Microbiome data curation</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="https://github.com/nfj1380/mrIML"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Graphical network model (GNM)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nfj1380/mrIML/blob/master/vignettes/Graphical%20network%20model%20(GNM).Rmd"><code>vignettes/Graphical network model (GNM).Rmd</code></a></small>
      <div class="d-none name"><code>Graphical network model (GNM).Rmd</code></div>
    </div>

    
    
<p>This vignette guides you through the new functionality that turns
mrIML into a graphical network model (GNMs). There is also new functions
in MrIML 2.0 that can be used in non-GN models, such as for single taxon
SDMs, epi models, gene environment association (GEA) studies and for
landscape genetics. Advances in Tidymodels are in the core of the
functions as are exciting advances in interpretable machine learning
(e.g., for quantifying interactions quickly). The capability to capture
uncertainty via bootstraps and spatial patterns are also a feature of
this MRIML module. The examples are from the microbiome world but of
course generalizable to any multi-response problem.</p>
<p>Currently, the models are optimized and tested on presence/absence
data. Models suitable for abundance data are being added in the
future.</p>
<p>First, we need to load up some extra functionality and some data. The
first example is coinfection data from New Caledonian <em>Zosterops</em>
species. A single continuous covariate is also included
(scale.prop.zos), which reflects the relative abundance of
<em>Zosterops</em> species among different sample sites.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Bird.parasites</span>,</span>
<span>                   <span class="op">-</span><span class="va">scale.prop.zos</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="co">#response variables eg. SNPs, pathogens, species....</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Bird.parasites</span>,</span>
<span>                   <span class="va">scale.prop.zos</span><span class="op">)</span> <span class="co"># feature set</span></span>
<span></span>
<span><span class="va">X1</span> <span class="op">&lt;-</span> <span class="va">Y</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">X1_fact</span> <span class="op">&lt;-</span> <span class="va">X1</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_all</a></span><span class="op">(</span><span class="va">as.factor</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_all</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">.</span> <span class="op">==</span> <span class="fl">0</span>, <span class="st">"absent"</span>, <span class="st">"present"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Note that the inclusion of ‘X1’ converts MrIML into a JSDM by
directly adding of the presence/absence patterns of the other taxa into
the model (as well as environmental/host covariates). Note, the order of
X1 and Y need to match.</p>
<div class="section level2">
<h2 id="setting-up-the-models">Setting up the models<a class="anchor" aria-label="anchor" href="#setting-up-the-models"></a>
</h2>
<p>Built into the MrIML architecture (and the big advantage with
Tidymodels) is the capability to change the underlying model easily. We
are going to set up two models to compare: a random forest model (RF)
and a logistic regression (lm). MrIML takes advantage of multi-core
processing so we set that up here to run on 5 cores. These steps are the
same as in MrIML 1.0.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_rf</span> <span class="op">&lt;-</span><span class="fu">rand_forest</span><span class="op">(</span>trees <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                       mode <span class="op">=</span> <span class="st">"classification"</span>,</span>
<span>                       mtry <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>,</span>
<span>                       min_n <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co">#100 trees are set for brevity. Aim to start with 1000</span></span>
<span>          <span class="fu">set_engine</span><span class="op">(</span><span class="st">"randomForest"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">model_lm</span> <span class="op">&lt;-</span> <span class="co">#model used to generate yhat</span></span>
<span>  <span class="fu">logistic_reg</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">set_engine</span><span class="op">(</span><span class="st">"glm"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">set_mode</span><span class="op">(</span><span class="st">"classification"</span><span class="op">)</span> <span class="co">#just for your response</span></span>
<span></span>
<span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu">plan</span><span class="op">(</span><span class="va">cluster</span>, workers<span class="op">=</span><span class="va">cl</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-the-models">Running the models<a class="anchor" aria-label="anchor" href="#running-the-models"></a>
</h2>
<p>Aside from adding JSDM functionality with the X1 call, we have also
enabled MrIML to tune hyperparameters using the very efficient ‘racing’
option (see Kuhn (2014)). In brief, this racing option take a small
subsample of parameters and eliminates parameter combinations that do
not improve fit using a repeated measure ANOVA model. Turning racing to
‘FALSE’ goes back to a grid search if you want to manually set the
tuning grid size. For the logistic regression there are no parameters to
tune so set to ‘FALSE’.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#random forest</span></span>
<span><span class="va">yhats_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">X</span>, Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                          X1<span class="op">=</span><span class="va">X1</span>,</span>
<span>                          Model<span class="op">=</span><span class="va">model_rf</span>,</span>
<span>                          balance_data<span class="op">=</span><span class="st">'no'</span>,</span>
<span>                          mode<span class="op">=</span><span class="st">'classification'</span>,</span>
<span>                          seed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample.int</a></span><span class="op">(</span><span class="fl">1e8</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                          morans<span class="op">=</span><span class="cn">F</span>,</span>
<span>                          prop<span class="op">=</span><span class="fl">0.7</span>, k<span class="op">=</span><span class="fl">5</span>, racing<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |==================                                                    |  25%  |                                                                              |===================================                                   |  50%  |                                                                              |====================================================                  |  75%  |                                                                              |======================================================================| 100%</span></span>
<span></span>
<span><span class="co">#linear model</span></span>
<span><span class="va">yhats_lm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">X</span>,Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                            X1<span class="op">=</span><span class="va">X1_fact</span>,</span>
<span>                            Model<span class="op">=</span><span class="va">model_lm</span> , </span>
<span>                            balance_data<span class="op">=</span><span class="st">'no'</span>,</span>
<span>                            mode<span class="op">=</span><span class="st">'classification'</span>,</span>
<span>                            seed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample.int</a></span><span class="op">(</span><span class="fl">1e8</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                            prop<span class="op">=</span><span class="fl">0.6</span>, racing<span class="op">=</span><span class="cn">F</span>, k<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |==================                                                    |  25%</span></span>
<span><span class="co">#&gt; Warning: No tuning parameters have been detected, performance will be evaluated</span></span>
<span><span class="co">#&gt; using the resamples with no tuning. Did you want to [tune()] parameters?</span></span>
<span><span class="co">#&gt;   |                                                                              |===================================                                   |  50%</span></span>
<span><span class="co">#&gt; Warning: No tuning parameters have been detected, performance will be evaluated</span></span>
<span><span class="co">#&gt; using the resamples with no tuning. Did you want to [tune()] parameters?</span></span>
<span><span class="co">#&gt;   |                                                                              |====================================================                  |  75%</span></span>
<span><span class="co">#&gt; Warning: No tuning parameters have been detected, performance will be evaluated</span></span>
<span><span class="co">#&gt; using the resamples with no tuning. Did you want to [tune()] parameters?</span></span>
<span><span class="co">#&gt;   |                                                                              |======================================================================| 100%</span></span>
<span><span class="co">#&gt; Warning: No tuning parameters have been detected, performance will be evaluated</span></span>
<span><span class="co">#&gt; using the resamples with no tuning. Did you want to [tune()] parameters?</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="comparing-performance">Comparing performance<a class="anchor" aria-label="anchor" href="#comparing-performance"></a>
</h2>
<p>It’s important to compare if there are any advantages to using a
random forest approach. Interpreation would be easier overall if
logistic regression gave similar predictive performance.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ModelPerf_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_rf</span>,</span>
<span>                                 Model<span class="op">=</span><span class="va">model_rf</span>,</span>
<span>                                 Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                                 mode<span class="op">=</span><span class="st">'classification'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ModelPerf_rf</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co">#across all parasites</span></span>
<span><span class="co">#&gt;       response  model_name           roc_AUC               mcc</span></span>
<span><span class="co">#&gt; 1   Hkillangoi rand_forest 0.886292016806723 0.365745522166111</span></span>
<span><span class="co">#&gt; 2  Hzosteropis rand_forest 0.941444372153546 0.718241208825566</span></span>
<span><span class="co">#&gt; 3 Microfilaria rand_forest 0.928861788617886 0.437856895997608</span></span>
<span><span class="co">#&gt; 4         Plas rand_forest 0.864678899082569 0.496552216869595</span></span>
<span><span class="co">#&gt;         sensitivity               ppv       specificity         prevalence</span></span>
<span><span class="co">#&gt; 1 0.983193277310924              0.25 0.906976744186047  0.115812917594655</span></span>
<span><span class="co">#&gt; 2 0.933962264150943 0.793103448275862 0.942857142857143  0.265033407572383</span></span>
<span><span class="co">#&gt; 3 0.983739837398374 0.333333333333333 0.937984496124031 0.0979955456570156</span></span>
<span><span class="co">#&gt; 4 0.926605504587156 0.538461538461538 0.893805309734513  0.195991091314031</span></span>
<span><span class="va">ModelPerf_rf</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="co">#overall</span></span>
<span><span class="co">#&gt; [1] 0.504599</span></span>
<span></span>
<span><span class="va">ModelPerf_lm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_lm</span>,</span>
<span>                                 Model<span class="op">=</span><span class="va">model_lm</span>,</span>
<span>                                 Y<span class="op">=</span><span class="va">Y</span>, mode<span class="op">=</span><span class="st">'classification'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ModelPerf_lm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt;       response   model_name           roc_AUC               mcc</span></span>
<span><span class="co">#&gt; 1   Hkillangoi logistic_reg 0.803326474622771              &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2  Hzosteropis logistic_reg 0.815781274341798   0.4040224605406</span></span>
<span><span class="co">#&gt; 3 Microfilaria logistic_reg  0.91820987654321 0.433340830644394</span></span>
<span><span class="co">#&gt; 4         Plas logistic_reg 0.777072192513369 0.240820814553862</span></span>
<span><span class="co">#&gt;         sensitivity               ppv       specificity         prevalence</span></span>
<span><span class="co">#&gt; 1                 1                 0               0.9  0.115812917594655</span></span>
<span><span class="co">#&gt; 2 0.946564885496183  0.36734693877551               0.8  0.265033407572383</span></span>
<span><span class="co">#&gt; 3 0.981481481481482 0.333333333333333 0.929824561403509 0.0979955456570156</span></span>
<span><span class="co">#&gt; 4 0.963235294117647 0.181818181818182 0.784431137724551  0.195991091314031</span></span>
<span><span class="va">ModelPerf_lm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 0.269546</span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrPerformancePlot.html">mrPerformancePlot</a></span><span class="op">(</span>ModelPerf1<span class="op">=</span><span class="va">ModelPerf_lm</span>,</span>
<span>                           ModelPerf2<span class="op">=</span><span class="va">ModelPerf_rf</span>,</span>
<span>                           mod_names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'linear_reg'</span>,<span class="st">'rand_forest'</span><span class="op">)</span>,</span>
<span>                           mode<span class="op">=</span><span class="st">'classification'</span> <span class="op">)</span> </span>
<span><span class="va">plots</span></span>
<span><span class="co">#&gt; [[1]]</span></span></code></pre></div>
<p><img src="Graphical%20network%20model%20(GNM)_files/figure-html/unnamed-chunk-4-1.png" width="768" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span></code></pre>
<p><img src="Graphical%20network%20model%20(GNM)_files/figure-html/unnamed-chunk-4-2.png" width="768" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 5</span></span></span>
<span><span class="co">#&gt;   response     outlier logistic_reg rand_forest diff_mod1_2</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Hkillangoi   <span style="color: #BB0000;">NA</span>             0           0.366     0.366  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Hzosteropis  <span style="color: #BB0000;">NA</span>             0.404       0.718     0.314  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Microfilaria <span style="color: #BB0000;">NA</span>             0.433       0.438     0.004<span style="text-decoration: underline;">52</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Plas         <span style="color: #BB0000;">NA</span>             0.241       0.497     0.256</span></span></code></pre>
<p>If we just look at the overall AUC values, it looks like model
performance is quite similar (0.86 for the rf and 0.81 for the lm).
However, when we look a the Mathew’s correlation coefficient (MCC) for
each taxa in the lm model we see that for <em>H.killangoi</em> and Plas
(<em>Plasmodium</em>) the values are much lower (e.g. 0.02, basically
just a guess) compared to 0.31 for rf). Remember the classes are
imbalanced so AUC tends to be an inflated measure. This is evidence that
non-linear relationships can make a difference to prediction overall. We
will interrogate the rf model further.</p>
<p>We will now ask if including putative associations between taxa
improves model performance overall or in the relationship between the
parasites or is host relative abundance enough.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">yhats_rf_noAssoc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">X</span>,</span>
<span>                                  Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                                  X1<span class="op">=</span><span class="cn">NULL</span>, <span class="co">#no associations for this one</span></span>
<span>                                  Model<span class="op">=</span><span class="va">model_rf</span>,</span>
<span>                                  balance_data<span class="op">=</span><span class="st">'no'</span>,</span>
<span>                                  mode<span class="op">=</span><span class="st">'classification'</span>,</span>
<span>                                  seed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample.int</a></span><span class="op">(</span><span class="fl">1e8</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                  prop<span class="op">=</span><span class="fl">0.7</span>,</span>
<span>                                  k<span class="op">=</span><span class="fl">5</span>,</span>
<span>                                  racing<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |==================                                                    |  25%  |                                                                              |===================================                                   |  50%  |                                                                              |====================================================                  |  75%  |                                                                              |======================================================================| 100%</span></span>
<span></span>
<span><span class="va">ModelPerf_rf_noAssoc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_rf_noAssoc</span>,</span>
<span>                                         Model<span class="op">=</span><span class="va">model_rf</span>,</span>
<span>                                         Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                                         mode<span class="op">=</span><span class="st">'classification'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ModelPerf_rf_noAssoc</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt;       response  model_name           roc_AUC               mcc</span></span>
<span><span class="co">#&gt; 1   Hkillangoi rand_forest  0.61633428300095              &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2  Hzosteropis rand_forest 0.787623248572911 0.494420277562475</span></span>
<span><span class="co">#&gt; 3 Microfilaria rand_forest 0.715175953079179 0.371709945158283</span></span>
<span><span class="co">#&gt; 4         Plas rand_forest  0.78360768175583 0.460220713518774</span></span>
<span><span class="co">#&gt;         sensitivity               ppv       specificity         prevalence</span></span>
<span><span class="co">#&gt; 1                 1                 0 0.866666666666667  0.115812917594655</span></span>
<span><span class="co">#&gt; 2 0.936170212765957  0.48780487804878 0.807339449541284  0.265033407572383</span></span>
<span><span class="co">#&gt; 3 0.983870967741935 0.272727272727273 0.938461538461538 0.0979955456570156</span></span>
<span><span class="co">#&gt; 4 0.898148148148148 0.555555555555556 0.889908256880734  0.195991091314031</span></span>
<span><span class="va">ModelPerf_rf</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co">#performance including associations</span></span>
<span><span class="co">#&gt;       response  model_name           roc_AUC               mcc</span></span>
<span><span class="co">#&gt; 1   Hkillangoi rand_forest 0.886292016806723 0.365745522166111</span></span>
<span><span class="co">#&gt; 2  Hzosteropis rand_forest 0.941444372153546 0.718241208825566</span></span>
<span><span class="co">#&gt; 3 Microfilaria rand_forest 0.928861788617886 0.437856895997608</span></span>
<span><span class="co">#&gt; 4         Plas rand_forest 0.864678899082569 0.496552216869595</span></span>
<span><span class="co">#&gt;         sensitivity               ppv       specificity         prevalence</span></span>
<span><span class="co">#&gt; 1 0.983193277310924              0.25 0.906976744186047  0.115812917594655</span></span>
<span><span class="co">#&gt; 2 0.933962264150943 0.793103448275862 0.942857142857143  0.265033407572383</span></span>
<span><span class="co">#&gt; 3 0.983739837398374 0.333333333333333 0.937984496124031 0.0979955456570156</span></span>
<span><span class="co">#&gt; 4 0.926605504587156 0.538461538461538 0.893805309734513  0.195991091314031</span></span></code></pre></div>
<p>You can see that overall including associations improved model
performance overall but particularly in predicting <em>H.killangoi</em>
and <em>Microfilaria</em>. Using MCC to compare models is problematic as
for the association free model its not defined for these taxa (NA,
probably as there are no false negatives for the low prevalence taxa in
the association-free model). Positive predictive value (PPV) is useful
in this case and shows that without associations we can’t predict the
occurrence of these taxa (PPV=0 for both). Including associations
increases PPV to ~0.2 - not great as 80% of our positive predictions for
these taxa are false.</p>
</div>
<div class="section level2">
<h2 id="downsampling">Downsampling<a class="anchor" aria-label="anchor" href="#downsampling"></a>
</h2>
<p>Including associations makes a difference, but how can we do better
in predicting our two rarer taxa? Up sampling is possible but in this
case we’ll try down sampling to see if correcting for class imbalance
improves our model fit.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">yhats_rf_downSamp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">X</span>,</span>
<span>                                   Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                                   X1<span class="op">=</span><span class="va">X1</span>,</span>
<span>                                   Model<span class="op">=</span><span class="va">model_rf</span>,</span>
<span>                                   balance_data<span class="op">=</span><span class="st">'down'</span>, <span class="co">#down sampling</span></span>
<span>                                   mode<span class="op">=</span><span class="st">'classification'</span>,</span>
<span>                                   seed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample.int</a></span><span class="op">(</span><span class="fl">1e8</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                   prop<span class="op">=</span><span class="fl">0.75</span>,</span>
<span>                                   k<span class="op">=</span><span class="fl">5</span>,</span>
<span>                                   racing<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |==================                                                    |  25%  |                                                                              |===================================                                   |  50%  |                                                                              |====================================================                  |  75%  |                                                                              |======================================================================| 100%</span></span>
<span></span>
<span><span class="va">ModelPerf_rf_downSamp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_rf_downSamp</span>,</span>
<span>                                          Model<span class="op">=</span><span class="va">model_rf</span>,</span>
<span>                                          Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                                          mode<span class="op">=</span><span class="st">'classification'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ModelPerf_rf_downSamp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt;       response  model_name           roc_AUC               mcc</span></span>
<span><span class="co">#&gt; 1   Hkillangoi rand_forest 0.893041237113402 0.489868155711274</span></span>
<span><span class="co">#&gt; 2  Hzosteropis rand_forest 0.915567765567766 0.645516132718224</span></span>
<span><span class="co">#&gt; 3 Microfilaria rand_forest  0.96504854368932 0.474958292677348</span></span>
<span><span class="co">#&gt; 4         Plas rand_forest  0.83719806763285 0.400395061824689</span></span>
<span><span class="co">#&gt;         sensitivity               ppv       specificity         prevalence</span></span>
<span><span class="co">#&gt; 1 0.814432989690722            0.8125 0.963414634146341  0.115812917594655</span></span>
<span><span class="co">#&gt; 2 0.897435897435897 0.742857142857143 0.886075949367089  0.265033407572383</span></span>
<span><span class="co">#&gt; 3 0.766990291262136                 1                 1 0.0979955456570156</span></span>
<span><span class="co">#&gt; 4               0.8 0.652173913043478               0.9  0.195991091314031</span></span></code></pre></div>
<p>Look at those PPV values now - much better. Our false positive rate
is down to ~ &lt; 15% overall. Now that we are happy with the
performance of the model, now we can interrogate further.</p>
</div>
<div class="section level2">
<h2 id="interpreting-the-model">Interpreting the model<a class="anchor" aria-label="anchor" href="#interpreting-the-model"></a>
</h2>
<p>In many cases, like this data set, community or microbiome data tend
to be small in size. When we apply stochastic machine learning
algorithms to such data, it can lead to challenges. For instance, the
importance of variables may vary substantially when we create multiple
models using the same data and algorithm. To handle this variability and
better understand prediction uncertainty, MrIML 2.0 has functionality to
capture uncertainty in our tuned model using bootstraps. Additionally,
this approach helps us estimate how variables affect the response, and
these estimates align with the results obtained from traditional linear
regression models (see Cook et al., 2021).</p>
<p>MrIML2.0 makes it easy to get bootstrap estimates for a variety of
interpretable machine learning tools, and uses these estimates to
construct marginalized co-occurrence networks. First, lets do the
bootstrapping and calculate variable importance.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> <span class="co">#can increase the number of cores as needed.</span></span>
<span><span class="fu">plan</span><span class="op">(</span><span class="va">cluster</span>, workers<span class="op">=</span><span class="va">cl</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#do bootstraps.</span></span>
<span><span class="va">bs_malaria</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrBootstrap.html">mrBootstrap</a></span><span class="op">(</span>yhats<span class="op">=</span><span class="va">yhats_rf</span>,</span>
<span>                          Y<span class="op">=</span><span class="va">Y</span>, </span>
<span>                          num_bootstrap <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                          downsample <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          mode<span class="op">=</span><span class="st">'classification'</span><span class="op">)</span> </span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |==================                                                    |  25%  |                                                                              |===================================                                   |  50%  |                                                                              |====================================================                  |  75%  |                                                                              |======================================================================| 100%</span></span>
<span><span class="co">#make sure downsample=TRUE as this did improve performance</span></span>
<span><span class="co">#just 10 bootstraps to keep this short. We suggest using more for a final analysis (100 is reasonable but depends on how big the data is)</span></span>
<span></span>
<span><span class="co">#up to here -not working properly</span></span>
<span><span class="va">bs_impVI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrVip.html">mrvip</a></span><span class="op">(</span></span>
<span>  mrBootstrap_obj <span class="op">=</span> <span class="va">bs_malaria</span>,</span>
<span>  yhats <span class="op">=</span> <span class="va">yhats_rf_downSamp</span>,</span>
<span>  X <span class="op">=</span> <span class="va">X</span>,</span>
<span>  X1 <span class="op">=</span> <span class="va">X1</span>,</span>
<span>  Y <span class="op">=</span> <span class="va">Y</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">'classification'</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.0</span>,</span>
<span>  global_top_var <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  local_top_var <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  taxa <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ModelPerf <span class="op">=</span> <span class="va">ModelPerf_rf_downSamp</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "here"</span></span>
<span></span>
<span><span class="va">bs_impVI</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>  <span class="co">#importance plot. There are plenty of other insights possible</span></span></code></pre></div>
<p><img src="Graphical%20network%20model%20(GNM)_files/figure-html/unnamed-chunk-7-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#the 'global_top_var' provides a limit to how many predictors are included in the community-wide plot. 'local_top_var provides a limit to the number of individual taxa plots. The threshold excludes plotting individual importance plots for taxa not well predicted by the model.</span></span></code></pre></div>
<p>You can see that host abundance is the most important predictor of
this parasite community (followed by <em>H.zosteropsis</em>) . However,
the second figure shows that there is important variability. For
example, <em>H.zosteropsis</em> is the most important predictor for the
occurence of <em>Microfilaria</em>, and host abundance (shortened to
sc..) is less important.</p>
</div>
<div class="section level2">
<h2 id="bootstrap-partial-dependence-plots">Bootstrap partial dependence plots<a class="anchor" aria-label="anchor" href="#bootstrap-partial-dependence-plots"></a>
</h2>
<p>To look at the the relationship between each variable and community
structure, MrIML 2.0 has a convenient wrapper to plot bootsrapped
partial dependencies for a taxa of interest.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrPD_bootstrap.html">mrPD_bootstrap</a></span><span class="op">(</span>mrBootstrap_obj<span class="op">=</span><span class="va">bs_malaria</span>,</span>
<span>                      vi_obj<span class="op">=</span><span class="va">bs_impVI</span>,</span>
<span>                      <span class="va">X</span>,</span>
<span>                      <span class="va">Y</span>,</span>
<span>                      target<span class="op">=</span><span class="st">'Plas'</span>,</span>
<span>                      global_top_var<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="Graphical%20network%20model%20(GNM)_files/figure-html/unnamed-chunk-8-1.png" width="768" style="display: block; margin: auto;">
These plots show that, for example, the presence of
<em>Microfilaria</em> greatly increass the probability of observing
<em>Plasmodium</em> (from ~0.32 to 0.58 while holding all other
variables at their mean value). Note that these are marginal
relationships (i.e. isolating the effect of just each predictor). When
host abundance is high the probability of detecting <em>Plasmodium</em>
descreases non-linearly (a threshold of around ~0). The other parasites
have less effect.</p>
<p>If we want to explore the effect of host abundance overall we can use
the ‘mr_Covar’ function.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">covar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_Covar.html">mr_Covar</a></span><span class="op">(</span>yhats<span class="op">=</span><span class="va">yhats_rf_downSamp</span>,</span>
<span>                  X<span class="op">=</span><span class="va">X</span>,</span>
<span>                  X1<span class="op">=</span><span class="va">X1</span>,</span>
<span>                  Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                  var<span class="op">=</span><span class="st">'scale.prop.zos'</span>,</span>
<span>                  sdthresh <span class="op">=</span><span class="fl">0.01</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Graphical%20network%20model%20(GNM)_files/figure-html/unnamed-chunk-9-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#sdthrsh just plots taxa responding the most.</span></span></code></pre></div>
<p>Note that this isn’t bootstrapped now - each line represents a taxon
in this case. The second plot shows the community-wide change in
occurrence probabilities across host abundance. Note the the occurrence
probabilities of all taxa drop an intermediate levels of host abundance
(0.75-1.25).</p>
</div>
<div class="section level2">
<h2 id="co-occurrence-network">Co-occurrence network<a class="anchor" aria-label="anchor" href="#co-occurrence-network"></a>
</h2>
<p>We can utilise all the bootstrapped partial dependence estimates
(pds) to construct a co-occurrence network. We show how this object can
be converted to an igraph object and plotted. This is a directed network
and edges are scaled by the standard deviation of the marginal change in
prediction. Red are positive associations (the predicted occurrence of
the taxa increases with the presence of the other) and blue are negative
(the predicted occurrence of the taxa decreases with the presence of the
other).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assoc_net</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrCoOccurNet_bootstrap.html">mrCoOccurNet_bootstrap</a></span> <span class="op">(</span>mrPD_obj<span class="op">=</span><span class="va">pds</span>,</span>
<span>                                    Y<span class="op">=</span><span class="va">Y</span><span class="op">)</span></span>
<span></span>
<span><span class="va">assoc_net_filtered</span> <span class="op">&lt;-</span>  <span class="va">assoc_net</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">mean_strength</span> <span class="op">&gt;</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="co">#based on our simulations the following rule of thumb for associations. Any association  &lt; 0.05  for mean strength is included.</span></span>
<span></span>
<span><span class="co">#convert to igraph</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">graph_from_data_frame</span><span class="op">(</span><span class="va">assoc_net_filtered</span>,</span>
<span>                           directed<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                           vertices<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span><span class="op">)</span> <span class="co">#matching Y data</span></span>
<span></span>
<span><span class="fu">E</span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">Value</span> <span class="op">&lt;-</span> <span class="va">assoc_net_filtered</span><span class="op">$</span><span class="va">mean_strength</span></span>
<span></span>
<span><span class="fu">E</span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">Color</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">assoc_net_filtered</span><span class="op">$</span><span class="va">direction</span> <span class="op">==</span> <span class="st">"negative"</span>, <span class="st">"blue"</span>, <span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert the igraph object to a ggplot object with NMDS layout</span></span>
<span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="fu">ggnetwork</span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the graph</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gg</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, xend <span class="op">=</span> <span class="va">xend</span>, yend <span class="op">=</span> <span class="va">yend</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_edges</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">Color</span>, linewidth <span class="op">=</span> <span class="op">(</span><span class="va">Value</span><span class="op">)</span><span class="op">)</span>, </span>
<span>             curvature <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>             arrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/arrow.html" class="external-link">arrow</a></span><span class="op">(</span>length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">5</span>, <span class="st">"pt"</span><span class="op">)</span>,</span>
<span>                           type <span class="op">=</span> <span class="st">"closed"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_nodes</span><span class="op">(</span>color <span class="op">=</span> <span class="st">"gray"</span>, size <span class="op">=</span> <span class="fu">degree</span><span class="op">(</span><span class="va">g</span>, mode <span class="op">=</span> <span class="st">"out"</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_identity.html" class="external-link">scale_color_identity</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu">geom_nodelabel_repel</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">name</span><span class="op">)</span>,</span>
<span>                       box.padding <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="st">"lines"</span><span class="op">)</span>,</span>
<span>                       data <span class="op">=</span> <span class="va">gg</span>,</span>
<span>                       size<span class="op">=</span><span class="fl">2</span>,</span>
<span>                       segment.colour <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                       colour <span class="op">=</span> <span class="st">"white"</span>, fill <span class="op">=</span> <span class="st">"grey36"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Graphical%20network%20model%20(GNM)_files/figure-html/unnamed-chunk-10-1.png" width="768" style="display: block; margin: auto;">
## 1 way, 2 way and 3 way interactions</p>
<p>Finally, we can quantify the importance of interactions overall as
well as one and two-way interaction importance using the same bootstrap
approach.See <a href="https://github.com/mayer79/hstats" class="external-link uri">https://github.com/mayer79/hstats</a> for more details about
the method.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">int_</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrInteractions.html">mrInteractions</a></span><span class="op">(</span>yhats<span class="op">=</span><span class="va">yhats_rf</span>,</span>
<span>                       <span class="va">X</span>,</span>
<span>                       <span class="va">Y</span>,</span>
<span>                       num_bootstrap<span class="op">=</span><span class="fl">10</span>,</span>
<span>                       feature <span class="op">=</span> <span class="st">'Plas'</span>,</span>
<span>                       top.int<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#10 bootstraps to keep it short. top int focusses on the 10 top interactions (all of them in this case).</span></span>
<span></span>
<span><span class="va">int_</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co"># overall plot</span></span></code></pre></div>
<p><img src="Graphical%20network%20model%20(GNM)_files/figure-html/unnamed-chunk-11-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">int_</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="co"># individual plot for the response of choice </span></span></code></pre></div>
<p><img src="Graphical%20network%20model%20(GNM)_files/figure-html/unnamed-chunk-11-2.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">int_</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="co">#two way plot</span></span></code></pre></div>
<p><img src="Graphical%20network%20model%20(GNM)_files/figure-html/unnamed-chunk-11-3.png" width="768" style="display: block; margin: auto;">
The first plot shows that interactions account for on average 27%
(bootstrap interval 23-34%) of variation in predictions for
<em>H.killangoi</em> and less for the other taxa. The second plot shows
that interactions involving host abundance impacts predictions of
<em>Plasmodium</em> the most (but <em>H.zosteropis</em> is also
important). This trend is similar community-wide. The next plot shows
that the interaction between <em>Haemoproteus</em> species is the most
important 2-way interaction for <em>Plasmodium</em> but this isn’t true
community wide as host abudance and <em>H.zosteropis</em> is the
strongest interaction overall. Taken together, we can see that
interactions between taxa are mediated by host abundance.</p>
<p>Finally, we can explore specific interactions in more detail using 2D
partial dependence plots. In case we choose the one of the more
important 2-way interactions impacting the probability of detetcting
<em>Plasmodium</em>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrFlashlight.html">mrFlashlight</a></span><span class="op">(</span>yhats<span class="op">=</span><span class="va">yhats_rf_downSamp</span>,</span>
<span>                   X<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span><span class="op">)</span>,</span>
<span>                   Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                   response <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>                   index<span class="op">=</span><span class="fl">4</span>,</span>
<span>                   mode<span class="op">=</span><span class="st">'classification'</span><span class="op">)</span> <span class="co">#index=4 selects Plasmodium</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">light_profile2d</span><span class="op">(</span><span class="va">fl</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scale.prop.zos"</span>,<span class="st">"Hzosteropis"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Graphical%20network%20model%20(GNM)_files/figure-html/unnamed-chunk-12-1.png" width="768" style="display: block; margin: auto;">
So if <em>H.zosteropsis</em> is not present and the relative abundance
of <em>Zosterops</em> species is low the probability of observing
<em>Plasmodium</em> is high (~&gt;0.7).</p>
<div class="section level3">
<h3 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<p>Cook et al., 2021: <a href="https://doi.org/10.18651/RWP2021-12" class="external-link uri">https://doi.org/10.18651/RWP2021-12</a> Kuhn (2014): <a href="https://doi.org/10.48550/arXiv.1405.6974" class="external-link uri">https://doi.org/10.48550/arXiv.1405.6974</a> Fountain-Jones
et al(2021): <a href="https://doi.org/10.1111/1755-0998.13495" class="external-link uri">https://doi.org/10.1111/1755-0998.13495</a></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nick Fountain-Jone, <a href="https://machado-lab.github.io/" class="external-link">Gustavo Machado</a>, Chris Kozakiewicz, Nick Clark.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
