<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Regression working example • mrIML</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Regression working example">
<meta property="og:description" content="mrIML">
<meta property="og:image" content="nfj1380.io/mrIML/reference/figures/logo.png">
<meta property="og:image:alt" content="Multivariate–multi-response–interpretable machine learning">
<meta name="twitter:card" content="summary">
<meta name="twitter:creator" content="@gustavoetal">
<meta name="twitter:site" content="@gustavoetal">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mrIML</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Vignette_LandscapeGenetics.html">Classification working example</a>
    </li>
    <li>
      <a href="../articles/Vignette_regression.html">Regression working example</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="https://github.com/nfj1380/mrIML/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Vignette_regression_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Regression working example</h1>
                        <h4 class="author">Nick Fountain-Jones, Gustavo Machado</h4>
            
            <h4 class="date">2021-04-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nfj1380/mrIML/blob/master/vignettes/Vignette_regression.Rmd"><code>vignettes/Vignette_regression.Rmd</code></a></small>
      <div class="hidden name"><code>Vignette_regression.Rmd</code></div>

    </div>

    
    
<p><img src="https://user-images.githubusercontent.com/33707823/88988817-531ad080-d31d-11ea-8d76-f1ad0506e405.png" width="200" height="200"></p>
<p>MrIML is a R package allows users to generate and interpret multi-response models (i.e., joint species distribution models) leveraging advances in data science and machine learning. MrIML couples the tidymodel infrastructure developed by Max Kuhn and colleagues with model agnostic interpretable machine learning tools to gain insights into multiple response data such as. As such mrIML is flexible and easily extendable allowing users to construct everything from simple linear models to tree-based methods for each response using the same syntax and same way to compare predictive performance. In this vignette we will guide you through how to apply this package to ecological genomics problems using the regression functionality of the package. This data set comes from Fitzpatrick et al 2014 who were examining adaptive genetic variation in relation to geography and climate adaptation (current and future) in balsam poplar(Populus balsamifera). See Ecology Letters, (2014) doi: 10.1111/ele.12376. In this paper they used the similar gradient forests routine (see Ellis et al 2012 Ecology), and we show that MrIML can not only provide more flexible model choice and interpretive capabilities, but can derive new insights into the relationship between climate and genetic variation. Further, we show that linear models of each loci have slightly greater predictive performance.</p>
<p>We focus on the adaptive SNP loci from GIGANTEA-5 (GI5) gene that has known links to stem development, plant circadian clock and light perception pathway. The data is the proportion of individuals in that population with that SNP loci.</p>
<div id="lets-load-that-data-" class="section level3">
<h3 class="hasAnchor">
<a href="#lets-load-that-data-" class="anchor"></a>Lets load that data.</h3>
</div>
<div id="running-the-analyis" class="section level3">
<h3 class="hasAnchor">
<a href="#running-the-analyis" class="anchor"></a>Running the analyis</h3>
<p>Performing the analysis is very similar to our classification example. Lets start with a constructing a linear model for this data set. We set Model 1 to a linear regression. See <a href="https://www.tidymodels.org/find/" class="uri">https://www.tidymodels.org/find/</a> for other regression model options Note that ‘mode’ must be regression and in MrIMLpredicts, model has to be set to ‘regression’.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">model1</span> <span class="op">&lt;-</span> <span class="co">#model used to generate yhat</span>
  <span class="co"># specify that the model is a random forest</span>
  <span class="fu"><a href="https://parsnip.tidymodels.org/reference/linear_reg.html">linear_reg</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># select the engine/package that underlies the model</span>
  <span class="fu"><a href="https://parsnip.tidymodels.org/reference/set_engine.html">set_engine</a></span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># choose either the continuous regression or binary classification mode</span>
  <span class="fu"><a href="https://parsnip.tidymodels.org/reference/set_args.html">set_mode</a></span><span class="op">(</span><span class="st">"regression"</span><span class="op">)</span>

<span class="va">yhats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">X</span>,Y<span class="op">=</span><span class="va">Y</span>, model1<span class="op">=</span><span class="va">model1</span>, balance_data<span class="op">=</span><span class="st">'no'</span>, model<span class="op">=</span><span class="st">'regression'</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span>, seed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span><span class="op">(</span><span class="fl">1e8</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="co">## Balanced data= up updamples and down downsampled to create a balanced set. For regression 'no' has to be selected.</span>

<span class="co">#save(yhats, file='Regression_lm') #always a good idea</span></code></pre></div>
<p>Model performance can be examined the same way as in the classification example, however the metrics are different. We provide root mean square error (rmse) and R2. You can see that the overall R2 is 0.13 but there is substantial variation across loci in predictive performance.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ModelPerf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats</span>, <span class="va">model1</span>, X<span class="op">=</span><span class="va">X</span>, model<span class="op">=</span><span class="st">'regression'</span><span class="op">)</span>
<span class="va">ModelPerf</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co">#predictive performance for individual responses. </span>
<span class="co">#&gt;              response model_name       rmse     rsquared</span>
<span class="co">#&gt; 1   CANDIDATE_GI5_108 linear_reg 0.04796608 2.546891e-03</span>
<span class="co">#&gt; 2   CANDIDATE_GI5_198 linear_reg 0.15048866 7.624520e-03</span>
<span class="co">#&gt; 3   CANDIDATE_GI5_268 linear_reg 0.10029507 2.906169e-03</span>
<span class="co">#&gt; 4    CANDIDATE_GI5_92 linear_reg 0.06612052 4.426335e-01</span>
<span class="co">#&gt; 5  CANDIDATE_GI5_1950 linear_reg 0.17410673 1.649338e-01</span>
<span class="co">#&gt; 6  CANDIDATE_GI5_2382 linear_reg 0.12773130 1.013087e-01</span>
<span class="co">#&gt; 7  CANDIDATE_GI5_2405 linear_reg 0.14193054 1.393440e-01</span>
<span class="co">#&gt; 8  CANDIDATE_GI5_2612 linear_reg 0.21260792 5.127816e-03</span>
<span class="co">#&gt; 9  CANDIDATE_GI5_2641 linear_reg 0.06219554 4.857684e-01</span>
<span class="co">#&gt; 10   CANDIDATE_GI5_33 linear_reg 0.26522077 1.688147e-02</span>
<span class="co">#&gt; 11 CANDIDATE_GI5_3966 linear_reg 0.13611469 2.883780e-01</span>
<span class="co">#&gt; 12 CANDIDATE_GI5_5033 linear_reg 0.05749290 2.183099e-01</span>
<span class="co">#&gt; 13 CANDIDATE_GI5_5090 linear_reg 0.14208168 5.904861e-03</span>
<span class="co">#&gt; 14 CANDIDATE_GI5_5119 linear_reg 0.07584565 5.999729e-02</span>
<span class="co">#&gt; 15 CANDIDATE_GI5_8997 linear_reg 0.14308321 5.526379e-05</span>
<span class="co">#&gt; 16 CANDIDATE_GI5_9287 linear_reg 0.08526388 2.076440e-01</span>
<span class="co">#&gt; 17 CANDIDATE_GI5_9447 linear_reg 0.11242858 1.149345e-01</span>
<span class="co">#&gt; 18 CANDIDATE_GI5_9551 linear_reg 0.14877236 8.588962e-02</span>
<span class="co">#&gt; 19 CANDIDATE_GI5_9585 linear_reg 0.16126089 3.030244e-01</span>
<span class="co">#&gt; 20 CANDIDATE_GI5_9659 linear_reg 0.16869780 5.816634e-02</span>
<span class="va">ModelPerf</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="co">#overall average r2 </span>
<span class="co">#&gt; [1] 0.1289852</span>

<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">ModelPerf</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="co">#save as a dataframe to compare to other models.</span></code></pre></div>
<p>Lets compare the performance of linear models to that of random forests. Random forests is the computational engine in gradient forests. Notice for random forests we have two hyperparamters to tune; mtry (number of features to randomly include at each split) and min_n (the minimum number of data points in a node that are required for the node to be split further). The syntax ‘tune()’ acts a placeholder to tell MrIML to tune those hyperparamters across a grid of values (defined in MRIML predicts ‘tune_grid_size’ argument). Different algorithms will have different hyperparameters.See <a href="https://www.tidymodels.org/find/parsnip/" class="uri">https://www.tidymodels.org/find/parsnip/</a> for parameter details. Note that large grid sizes (&gt;10) for algorithms with lots of hyperparameters (such as extreme gradient boosting) will be computationally demanding. In this case we choose a grid size of 5.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">model1</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://parsnip.tidymodels.org/reference/rand_forest.html">rand_forest</a></span><span class="op">(</span>trees <span class="op">=</span> <span class="fl">100</span>, mtry<span class="op">=</span><span class="fu">tune</span><span class="op">(</span><span class="op">)</span>, min_n <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>, mode <span class="op">=</span> <span class="st">"regression"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://parsnip.tidymodels.org/reference/set_engine.html">set_engine</a></span><span class="op">(</span><span class="st">"ranger"</span>, importance <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"impurity"</span>,<span class="st">"impurity_corrected"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://parsnip.tidymodels.org/reference/set_args.html">set_mode</a></span><span class="op">(</span><span class="st">"regression"</span><span class="op">)</span>

<span class="va">yhats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">X</span>,Y<span class="op">=</span><span class="va">Y</span>, model1<span class="op">=</span><span class="va">model1</span>, balance_data<span class="op">=</span><span class="st">'no'</span>, model<span class="op">=</span><span class="st">'regression'</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span>, tune_grid_size<span class="op">=</span><span class="fl">5</span> <span class="op">)</span>

<span class="co">#save(yhats, file='Regression_rf')</span>

<span class="va">ModelPerf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats</span>, <span class="va">model1</span>, X<span class="op">=</span><span class="va">X</span>, model<span class="op">=</span><span class="st">'regression'</span><span class="op">)</span>
<span class="va">ModelPerf</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co">#predictive performance for individual responses. </span>
<span class="co">#&gt;              response  model_name       rmse    rsquared</span>
<span class="co">#&gt; 1   CANDIDATE_GI5_108 rand_forest 0.04895187 0.119057391</span>
<span class="co">#&gt; 2   CANDIDATE_GI5_198 rand_forest 0.13756699 0.246909808</span>
<span class="co">#&gt; 3   CANDIDATE_GI5_268 rand_forest 0.05901889 0.456810346</span>
<span class="co">#&gt; 4    CANDIDATE_GI5_92 rand_forest 0.11419570 0.540519679</span>
<span class="co">#&gt; 5  CANDIDATE_GI5_1950 rand_forest 0.11909837 0.379185254</span>
<span class="co">#&gt; 6  CANDIDATE_GI5_2382 rand_forest 0.10249973 0.023551339</span>
<span class="co">#&gt; 7  CANDIDATE_GI5_2405 rand_forest 0.13926611 0.194581041</span>
<span class="co">#&gt; 8  CANDIDATE_GI5_2612 rand_forest 0.13235368 0.440702701</span>
<span class="co">#&gt; 9  CANDIDATE_GI5_2641 rand_forest 0.09328570          NA</span>
<span class="co">#&gt; 10   CANDIDATE_GI5_33 rand_forest 0.13287892 0.157436141</span>
<span class="co">#&gt; 11 CANDIDATE_GI5_3966 rand_forest 0.13649191 0.006767769</span>
<span class="co">#&gt; 12 CANDIDATE_GI5_5033 rand_forest 0.06748797          NA</span>
<span class="co">#&gt; 13 CANDIDATE_GI5_5090 rand_forest 0.08471521 0.457627185</span>
<span class="co">#&gt; 14 CANDIDATE_GI5_5119 rand_forest 0.05913244          NA</span>
<span class="co">#&gt; 15 CANDIDATE_GI5_8997 rand_forest 0.12459845 0.410904718</span>
<span class="co">#&gt; 16 CANDIDATE_GI5_9287 rand_forest 0.07828263 0.267496359</span>
<span class="co">#&gt; 17 CANDIDATE_GI5_9447 rand_forest 0.10409739 0.004125598</span>
<span class="co">#&gt; 18 CANDIDATE_GI5_9551 rand_forest 0.15230865 0.159312454</span>
<span class="co">#&gt; 19 CANDIDATE_GI5_9585 rand_forest 0.13619821 0.320886097</span>
<span class="co">#&gt; 20 CANDIDATE_GI5_9659 rand_forest 0.10806856 0.461709672</span>
<span class="va">ModelPerf</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="co">#overall average r2 </span>
<span class="co">#&gt; [1] 0.1065249</span>

<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">ModelPerf</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p>You can see that predictive performance is actually slightly less using random forests (overall R2 = 0.12) but for some loci random forests does better than our linear models and sometimes worse. Which to choose? Generally simpler models are preferred (the linear model in this case) but it depends on how important to think non-linear response are. In future versions of MrIML we will implement ensemble models that will overcome this issue. For the time-being we will have a look at variable importance for the random forest based model.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">VI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrVip.html">mrVip</a></span><span class="op">(</span><span class="va">yhats</span>, Y<span class="op">=</span><span class="va">Y</span><span class="op">)</span> 

<span class="fu"><a href="../reference/plot_vi.html">plot_vi</a></span><span class="op">(</span>VI<span class="op">=</span><span class="va">VI</span>,  X<span class="op">=</span><span class="va">X</span>,Y<span class="op">=</span><span class="va">Y</span>, modelPerf<span class="op">=</span><span class="va">ModelPerf</span>,  cutoff<span class="op">=</span> <span class="fl">0.1</span>, plot.pca<span class="op">=</span><span class="st">'yes'</span>, model<span class="op">=</span><span class="st">'regression'</span><span class="op">)</span></code></pre></div>
<p><img src="Vignette_regression_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<pre><code>#&gt; Press [enter] to plot individual variable importance summaries</code></pre>
<p><img src="Vignette_regression_files/figure-html/unnamed-chunk-4-2.png" width="700"></p>
<pre><code>#&gt; Press [enter] to plot the importance PCA plot</code></pre>
<p><img src="Vignette_regression_files/figure-html/unnamed-chunk-4-3.png" width="700"></p>
<p>Cutoff reduces the number of individual SNP plots presented in the second plot and ‘plot.pca=’yes’’ enables the variable importance scores to be analysed using principal component analysis (PCA) where SNPs closer in PCA space are shaped by similar combinations of features. You can see that bio_18 (summer precipitation), bio_1 (mean annual temperature) and bio_10 (mean summer temperature) are the most important features overall. Summer precipitation was not as important in Fitzpatrick et al but otherwise these results are similar. The second plot shows the individual models (with an r2 &gt; 0.1, for your data you will need to play around with this threshold) and you can see for some SNPs bio_1 is more important whereas for another MEM.1 is more prominent.The PCA shows that candidate 5119, 9287, 5033 and 108 are shaped similarly by the features we included and may, for example, be product of linked selection.</p>
<p>Now we can explore the model further my plotting the relationships between our SNPs and a feature in our set. Lets choose bio_1 (mean annual temperature) and plot the individual and global (average of all SNPs) partial dependency (PD) plots.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">flashlightObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrFlashlight.html">mrFlashlight</a></span><span class="op">(</span><span class="va">yhats</span>, <span class="va">X</span>, <span class="va">Y</span>, response <span class="op">=</span> <span class="st">"multi"</span>, model<span class="op">=</span><span class="st">'regression'</span><span class="op">)</span>

<span class="va">profileData_pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/flashlight/man/light_profile.html">light_profile</a></span><span class="op">(</span><span class="va">flashlightObj</span>, v <span class="op">=</span> <span class="st">"bio_1"</span><span class="op">)</span> <span class="co">#partial dependencies</span>

<span class="fu"><a href="../reference/mrProfileplot.html">mrProfileplot</a></span><span class="op">(</span><span class="va">profileData_pd</span> , sdthresh <span class="op">=</span><span class="fl">0.01</span><span class="op">)</span></code></pre></div>
<p><img src="Vignette_regression_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<pre><code>#&gt; `geom_smooth()` using formula 'y ~ x'</code></pre>
<p><img src="Vignette_regression_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
<p>The first plot is a partial dependency for all SNPs that respond to mean annual temperature. What we mean by respond here is that the prediction surface (the line) deviates across the Y axis of the PD plots. We measure this deviation by calculating the standard deviation and use that as a threshold (‘sd thresh=0.01’ in this case and this will differ by data set) to ease visualization of these relationships. The second plot is the smoothed average partial dependency of SNPs across a annual temperature gradient. This is very similar to the pattern observed by Fitzpatrick et al except with a slight decline in SNP turnover with mean annual temperatures &gt; 0. Combined,you can see here only few candidate SNPs are driving this pattern and these may warrant further interrogation.</p>
<p>Lets compare the PDs to accumulated local effect plots that are less sensitive to correlations among features (see Molnar 2019).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">profileData_ale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/flashlight/man/light_profile.html">light_profile</a></span><span class="op">(</span><span class="va">flashlightObj</span>, v <span class="op">=</span> <span class="st">"bio_1"</span>, type <span class="op">=</span> <span class="st">"ale"</span><span class="op">)</span> <span class="co">#accumulated local effects</span>

<span class="fu"><a href="../reference/mrProfileplot.html">mrProfileplot</a></span><span class="op">(</span><span class="va">profileData_ale</span> , sdthresh <span class="op">=</span><span class="fl">0.01</span><span class="op">)</span></code></pre></div>
<p><img src="Vignette_regression_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<pre><code>#&gt; `geom_smooth()` using formula 'y ~ x'</code></pre>
<p><img src="Vignette_regression_files/figure-html/unnamed-chunk-7-2.png" width="700"></p>
<p>The effect of mean annual temperature on SNP turnover is not as distinct in the global ALE plot. This may mean that correlations between features may be important for the predictions.</p>
<p>MrIML has easy to use functionality that can can quantify interactions between features. Note that this can take a while to compute.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">interactions</span> <span class="op">&lt;-</span><span class="fu"><a href="../reference/mrInteractions.html">mrInteractions</a></span><span class="op">(</span><span class="va">yhats</span>, <span class="va">X</span>, <span class="va">Y</span>,  model<span class="op">=</span><span class="st">'regression'</span><span class="op">)</span> <span class="co">#this is computationally intensive so multicores are needed.</span>

<span class="fu"><a href="../reference/mrPlot_interactions.html">mrPlot_interactions</a></span><span class="op">(</span><span class="va">interactions</span>, <span class="va">X</span>,<span class="va">Y</span>, top_ranking <span class="op">=</span> <span class="fl">2</span>, top_response<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="co">#can increase the number of interactions/SNPs ('responses') shown</span></code></pre></div>
<p><img src="Vignette_regression_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<pre><code>#&gt; Press [enter] to continue for response with strongest interactions</code></pre>
<p><img src="Vignette_regression_files/figure-html/unnamed-chunk-8-2.png" width="700"></p>
<pre><code>#&gt; Press [enter] to continue for individual response results</code></pre>
<p><img src="Vignette_regression_files/figure-html/unnamed-chunk-8-3.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">#Make sure you save your results</span>
<span class="co">#save(interactions, 'Fitzpatrick2016interactions')</span>

<span class="co">#load('Fitzpatrick2016interactions')</span></code></pre></div>
<p>The first plot reveals that the strongest global interaction is between mean annual (bio_1) temperature and summer precipitation (bio_10) but the interactions between other bioclim features are also relatively strong. Mean temperature interacts to some degree with spatial patterns as well (MEMs) to shape SNP turnover. Note that importance is all relative. the second plot shows that the predictions of candidate 33 are most effected by interacting features and the third plot shows that the interaction is between mean annual temperature and altitude.</p>
<p>This is touching only the surface of what is possible in terms of interrogating this model. See <a href="https://cran.r-project.org/web/packages/flashlight/vignettes/flashlight.html" class="uri">https://cran.r-project.org/web/packages/flashlight/vignettes/flashlight.html</a> for other options.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Nick Fountain-Jones, Nicholas Clark, Chris Kozakiewicz, <a href="https://gustavoetal.org/">Gustavo Machado</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
